### ISSUE TITLE
Questions about the technical report

### ISSUE THREAD

**Issue by StarCycle** [BAAI-DCAI/Bunny#8]
Hello,

It's a great work! And there are several questions:

1. In the technical report you mentioned 

> We find that LoRA empirically leads to better performance than fully tuning across all combinations of model architectures, probably because smaller models are more susceptible to catastrophic forgetting, while LoRA tuning alleviates this issue.

By " fully tuning across all combinations of model architectures", do you mean finetune the SigLIP encoder + projector + phi2, or just projector + phi2? And why LoRA tuning can alleviate catastrophic forgetting (sorry I am not familiar with this...)? Note that in [this paper](https://arxiv.org/abs/2309.10313), using LoRA cannot avoid the model overfitting the finetuning dataset.

2. How do you select the learning rate, batch size, and cosine annealing schedule? Do you perform hyperparameter searching?

To avoid overfitting, it seems that researchers only train LLaVA for one epoch (both the pretrain and finetuning phase). Therefore, the loss curve may not converge to the lowest point. 

For example, this is my loss curve and learning rate schedule during the pretrain phase:

![image](https://github.com/BAAI-DCAI/Bunny/assets/33491471/2b3c29df-d4d4-41c9-a984-aeb4cdadb94c)
![image](https://github.com/BAAI-DCAI/Bunny/assets/33491471/c782dd23-f265-4bd5-891e-a2659d0d6f4a)

And the loss curve and lr schedule in the finetune phase:

![image](https://github.com/BAAI-DCAI/Bunny/assets/33491471/7b4b3ca0-b53c-42ad-b12f-d47706df94a0)
![image](https://github.com/BAAI-DCAI/Bunny/assets/33491471/fcf17dfb-e737-47a4-ba99-a289ced1eab5)

I guess the network does not converge at all...so how do you determine these hyperparameters? Do you select a set of hyperparameters which makes your network fully converges? Or you just select a set of hyperparameters which has the best benchmark performance?

Best,
Starcycle

**Comment by StarCycle** [comment]
Btw it will be great if you can share the loss curve in the report!

I think the biggest problem of LLaVA based models is avoiding overfitting, since they only use limited training data to align vision encoder and llm

**Comment by Isaachhh** [comment]
Sorry for the late reply.

For Q1,
By  "fully tuning across all combinations of model architectures", we mean finetuning just projector + phi2.
And for the influence of LoRA, it's more a empirical finding that LoRA conveys better performance. You can see similar findings in [Idefics2 tech report](https://arxiv.org/abs/2405.02246).

For Q2,
We almost don't select a set of hyperparameters which makes the network fully converge. We conduct some experiments of learning rate and observe the loss curve and downstream performance.
BTW, the output loss of deepspeed is the **average** loss of this epoch. So, loss fluctuation around some level is normal given that the model sees every sample only once. Recently, some LLaVA variants conduct 2 epoch in sft stages ([mipha](https://arxiv.org/abs/2403.06199), [xtuner-phi3](https://huggingface.co/xtuner/llava-phi-3-mini-hf), [Prismatic VLMs](https://arxiv.org/abs/2402.07865)). If you train 2 epochs, you would see a loss jump in the begining of second epoch, due to the average loss output, too.

For the loss curve of Bunny, you can find the loss log in `trainer_state.json` in the related HuggingFace repo of each model.
For example, https://huggingface.co/BAAI/bunny-phi-2-siglip-lora/blob/main/trainer_state.json

**Comment by Isaachhh** [comment]
Close the issue for now if there's no further discussions. Feel free to reopen it if there's any other questions.
